ksql-engine/src/main/java/io/confluent/ksql/planner/plan/JoinNode.java

```diff
@@ -208,36 +230,197 @@ SchemaKTable tableForJoin(
+                                            leftStream.getKeyField().name()),
+                                 getSerDeForNode(joinNode.left));
+        case OUTER:
+          throw new KsqlException("Outer joins between streams and tables (stream: left, "
+                                  + "table: right) are not supported.");
+
+        default:
+          throw new KsqlException("Invalid join type encountered: " + joinNode.joinType);
+      }
+    }
+
+    SchemaKTable doTableToTableJoin() {
+      final SchemaKTable leftTable = buildTable(joinNode.getLeft());
+      final SchemaKTable rightTable = buildTable(joinNode.getRight());
+
+      switch (joinNode.joinType) {
+        case LEFT:
+          return leftTable.leftJoin(rightTable, joinNode.schema,
+                                    getJoinKey(joinNode.leftAlias, leftTable.getKeyField().name()));
+        case INNER:
+          return leftTable.join(rightTable, joinNode.schema,
+                                getJoinKey(joinNode.leftAlias, leftTable.getKeyField().name()));
+        case OUTER:
+          return leftTable.outerJoin(rightTable, joinNode.schema,
+                                     getJoinKey(joinNode.leftAlias,
+                                                leftTable.getKeyField().name()));
+        default:
+          throw new KsqlException("Invalid join type encountered: " + joinNode.joinType);
+      }
+
+    }
+
+    private Field getJoinKey(final String alias, final String keyFieldName) {
+      return joinNode.schema.field(alias + "." + keyFieldName);
+    }
+
+
+    private SchemaKStream buildStream(final PlanNode node, final String keyFieldName) {
+      return maybeRePartitionByKey(node.buildStream(builder, ksqlConfig, kafkaTopicClient,
+                                                    functionRegistry, props,
+                                                    schemaRegistryClient),
+                                   keyFieldName);
+    }
+
+    private SchemaKTable buildTable(final PlanNode node) {
+
+      Map<String, Object> joinTableProps = new HashMap<>(props);
+      joinTableProps.put(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest");
+
+      final SchemaKStream schemaKStream = node.buildStream(
+          builder,
+          ksqlConfig,
+          kafkaTopicClient,
+          functionRegistry,
+          joinTableProps, schemaRegistryClient);
+
+      if (!(schemaKStream instanceof SchemaKTable)) {
+        throw new KsqlException("Expected to find a Table, found a stream instead");
+      }
+
+      return (SchemaKTable) schemaKStream;
+    }
+
+    private SchemaKStream maybeRePartitionByKey(final SchemaKStream stream,
+                                                final String targetKey) {
+      final Schema schema = stream.getSchema();
+      final Field field =
+          SchemaUtil
+              .getFieldByName(schema,
+                              targetKey).orElseThrow(() -> new KsqlException("couldn't find "
```

Comment:
nit: I don't think `Schema` has a `toString` so this may not print anything helpful